### PR TITLE
Add methodForTest() convention to the style guide.

### PR DESCRIPTION
**Pull request by dnoe**
This naming convention is used throughout Envoy but I didn't see any
reference to it in the style guide.

### CODE REVIEW

File: STYLE.md [hunk 1]
```diff
@@ -52,6 +52,8 @@
   exceptions such as `main()` functions. When code cannot be placed inside the
   Envoy namespace there should be a comment of the form `// NOLINT(namespace-envoy)` at
   the top of the file.
+* If a public method is only intended to be called from test code then it should
+  have a name that ends in ForTest() such as aMethodForTest().
```
Comment (htuch) [inline STYLE.md]: Nit: ``ForTest()`` in quotes etc.

Comment (alyssawilk) [inline STYLE.md]: Thanks for adding this convention!

Is this also true for protected methods?  Would "If a method defined outside of the test directory is only intended..." be more clear or too wordy?

Comment (mattklein123) [inline STYLE.md]: Can we also add some kind of clause that this should be a last resort thing. IMO it's rare this is really necessary in prod code.

Comment (dnoe) [inline STYLE.md]: I didn't call it out here explicitly but my understanding is that in Envoy style this is preferable to doing something with `friend`-ing the test code. Thoughts on that?

Comment (dnoe) [inline STYLE.md]: I think that's a better way to say it, I'll update.

Comment (mattklein123) [inline STYLE.md]: I don't really have a strong opinion either way. My point mainly is that I think in practice it is actually quite rare that one needs to add a public method just for testing or friend the test code. Tests can frequently be written without doing any of that which IMO makes the tests better as the don't see into the inner details of the code under test. I would point that out, but I don't feel very strongly about it either way.
